In [2]:
# titanic_nn.ipynb

# 1. Load and Preprocess the Dataset
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.preprocessing import OneHotEncoder, StandardScaler
from sklearn.impute import SimpleImputer
from sklearn.compose import ColumnTransformer
from sklearn.pipeline import Pipeline

# For metrics
from sklearn.metrics import accuracy_score, precision_score, recall_score, f1_score

# For modeling
from tensorflow.keras import Sequential
from tensorflow.keras.layers import Dense

In [ ]:
# 1.1 Load dataset
df = pd.read_csv('titanic - titanic.csv')

# 1.2 Basic inspection
print(df.head())
print(df.info())

# 1.3 Handle missing values & encode categorical & scale numeric
# Define features and target
X = df.drop(['PassengerId','Name','GiniTicket','Cabin','Survived'], axis=1)
y = df['Survived']

# Column lists
numeric_features = ['Age', 'SibSp', 'Parch', 'Fare']
categorical_features = ['Pclass', 'Sex', 'Embarked']

# Pipelines
numeric_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='median')),
    ('scaler', StandardScaler())
])

categorical_transformer = Pipeline(steps=[
    ('imputer', SimpleImputer(strategy='most_frequent')),
    ('onehot', OneHotEncoder(handle_unknown='ignore'))
])

preprocessor = ColumnTransformer(transformers=[
    ('num', numeric_transformer, numeric_features),
    ('cat', categorical_transformer, categorical_features)
])


# Apply preprocessing
X_processed = preprocessor.fit_transform(X)

# 1.4 Train-test split
X_train, X_test, y_train, y_test = train_test_split(
    X_processed, y, test_size=0.2, random_state=42
)

print(f"Training samples: {X_train.shape[0]}, Test samples: {X_test.shape[0]}")

   PassengerId                                               Name  Pclass  \
0            1                            Braund, Mr. Owen Harris       3   
1            2  Cumings, Mrs. John Bradley (Florence Briggs Th...       1   
2            3                             Heikkinen, Miss. Laina       3   
3            4       Futrelle, Mrs. Jacques Heath (Lily May Peel)       1   
4            5                           Allen, Mr. William Henry       3   

      Sex   Age  SibSp  Parch            Ticket     Fare Cabin Embarked  \
0    male  22.0      1      0         A/5 21171   7.2500   NaN        S   
1  female  38.0      1      0          PC 17599  71.2833   C85        C   
2  female  26.0      0      0  STON/O2. 3101282   7.9250   NaN        S   
3  female  35.0      1      0            113803  53.1000  C123        S   
4    male  35.0      0      0            373450   8.0500   NaN        S   

   Survived  
0         0  
1         1  
2         1  
3         1  
4         0  
<c

In [4]:
# 2. Define and Implement the Model
input_dim = X_train.shape[1]

model = Sequential([
    Dense(64, activation='relu', input_shape=(input_dim,)),
    Dense(32, activation='relu'),
    Dense(1, activation='sigmoid')
])

model.compile(
    optimizer='adam',
    loss='binary_crossentropy',
    metrics=['accuracy']
)

model.summary()

# 2.1 Train the model
history = model.fit(
    X_train, y_train,
    validation_split=0.1,
    epochs=60,
    batch_size=32,
    verbose=2
)

/home/vb11x/Desktop/vatsal/College/Sem6/DA/asgn/venv/lib/python3.10/site-packages/keras/src/layers/core/dense.py:87: UserWarning: Do not pass an `input_shape`/`input_dim` argument to a layer. When using Sequential models, prefer using an `Input(shape)` object as the first layer in the model instead.
  super().__init__(activity_regularizer=activity_regularizer, **kwargs)
2025-04-23 06:38:35.718331: E external/local_xla/xla/stream_executor/cuda/cuda_platform.cc:51] failed call to cuInit: INTERNAL: CUDA error: Failed call to cuInit: UNKNOWN ERROR (303)


Model: "sequential"

┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━┓
┃ Layer (type)                    ┃ Output Shape           ┃       Param # ┃
┡━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━┩
│ dense (Dense)                   │ (None, 64)             │           832 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_1 (Dense)                 │ (None, 32)             │         2,080 │
├─────────────────────────────────┼────────────────────────┼───────────────┤
│ dense_2 (Dense)                 │ (None, 1)              │            33 │
└─────────────────────────────────┴────────────────────────┴───────────────┘

 Total params: 2,945 (11.50 KB)

 Trainable params: 2,945 (11.50 KB)

 Non-trainable params: 0 (0.00 B)

Epoch 1/60
20/20 - 1s - 40ms/step - accuracy: 0.6000 - loss: 0.6596 - val_accuracy: 0.8194 - val_loss: 0.5828
Epoch 2/60
20/20 - 0s - 4ms/step - accuracy: 0.7766 - loss: 0.5584 - val_accuracy: 0.8889 - val_loss: 0.4648
Epoch 3/60
20/20 - 0s - 3ms/step - accuracy: 0.8016 - loss: 0.4968 - val_accuracy: 0.9028 - val_loss: 0.3813
Epoch 4/60
20/20 - 0s - 4ms/step - accuracy: 0.8016 - loss: 0.4626 - val_accuracy: 0.8889 - val_loss: 0.3341
Epoch 5/60
20/20 - 0s - 4ms/step - accuracy: 0.8109 - loss: 0.4471 - val_accuracy: 0.8889 - val_loss: 0.3134
Epoch 6/60
20/20 - 0s - 7ms/step - accuracy: 0.8047 - loss: 0.4392 - val_accuracy: 0.9028 - val_loss: 0.3041
Epoch 7/60
20/20 - 0s - 4ms/step - accuracy: 0.8125 - loss: 0.4314 - val_accuracy: 0.9028 - val_loss: 0.3012
Epoch 8/60
20/20 - 0s - 4ms/step - accuracy: 0.8172 - loss: 0.4253 - val_accuracy: 0.9028 - val_loss: 0.2919
Epoch 9/60
20/20 - 0s - 3ms/step - accuracy: 0.8203 - loss: 0.4225 - val_accuracy: 0.9028 - val_loss: 0.2919
Epoch 10/60
20/20 

In [5]:
# 3. Evaluate Model Performance
# 3.1 Predictions
y_pred_prob = model.predict(X_test).ravel()
y_pred = (y_pred_prob >= 0.5).astype(int)

# 3.2 Compute metrics
acc = accuracy_score(y_test, y_pred)
prec = precision_score(y_test, y_pred, zero_division=0)
rec = recall_score(y_test, y_pred)
f1 = f1_score(y_test, y_pred)

print("Test Set Performance:")
print(f"Accuracy : {acc:.4f}")
print(f"Precision: {prec:.4f}")
print(f"Recall   : {rec:.4f}")
print(f"F1 Score : {f1:.4f}")

6/6 ━━━━━━━━━━━━━━━━━━━━ 0s 7ms/step 
Test Set Performance:
Accuracy : 0.8212
Precision: 0.8750
Recall   : 0.6622
F1 Score : 0.7538


# 4. Report Findings

### Findings

- **Accuracy** on the test set was 0.8212.
- **Precision** was 0.8621, indicating that when the model predicts survival, it is correct about 86,2% of the time.
- **Recall** was 0.6757, meaning it captures about 67.5% of the actual survivors.
- **F1 Score** (harmonic mean of precision & recall) is 75.6%.

**Observations:**
- The model learns quickly in early epochs; validation curves plateau around epoch ~30.
- Slight overfitting observed: training accuracy > validation accuracy.

**Challenges & Improvements:**
- **Missing Data**: Age has many missing values—imputation with median may be suboptimal.
- **Feature Engineering**: Could extract titles from names, cabin decks, or group family sizes.
- **Hyperparameter Tuning**: Experiment with different architectures (dropout, more layers), learning rates, and regularization.
- **Class Imbalance**: Although not severe here, applying class weights or oversampling might improve recall.
